packages/cheqd-sdk/src/anoncreds/services/CheqdSdkAnonCredsRegistry.ts

```diff
@@ -0,0 +1,300 @@
+
+import { v4 } from 'uuid'
+
+import { CheqdDidResolver, CheqdDidRegistrar } from '../../dids'
+import { cheqdSdkAnonCredsRegistryIdentifierRegex, parseCheqdDid } from '../utils/identifiers'
+import { Convert } from '../utils/transform'
+
+export class CheqdSdkAnonCredsRegistry implements AnonCredsRegistry {
+  /**
+   * This class supports resolving and registering objects with cheqd identifiers.
+   * It needs to include support for the schema, credential definition, revocation registry as well
+   * as the issuer id (which is needed when registering objects).
+   */
+  public readonly supportedIdentifier = cheqdSdkAnonCredsRegistryIdentifierRegex
+
+  public async getSchema(agentContext: AgentContext, schemaId: string): Promise<GetSchemaReturn> {
+    try {
+      const cheqdDidResolver = agentContext.dependencyManager.resolve(CheqdDidResolver)
+      const parsedDid = parseCheqdDid(schemaId)
+      if (!parsedDid) {
+        throw new Error(`Invalid schemaId: ${schemaId}`)
+      }
+
+      agentContext.config.logger.trace(`Submitting get schema request for schema '${schemaId}' to ledger`)
+
+      const response = await cheqdDidResolver.resolveResource(agentContext, schemaId)
+      const schema = Convert.toCheqdSchema(JSON.stringify(response.resource))
+
+      return {
+        schema: {
+          attrNames: schema.attrNames,
+          name: schema.name,
+          version: schema.version,
+          issuerId: parsedDid.did,
+        },
+        schemaId,
+        resolutionMetadata: {},
+        schemaMetadata: {},
+      }
+    } catch (error) {
+      agentContext.config.logger.error(`Error retrieving schema '${schemaId}'`, {
+        error,
+        schemaId,
+      })
+
+      return {
+        schemaId,
+        resolutionMetadata: {
+          error: 'notFound',
+          message: `unable to resolve schema: ${error.message}`,
+        },
+        schemaMetadata: {},
+      }
+    }
+  }
+
+  public async registerSchema(
+    agentContext: AgentContext,
+    options: RegisterSchemaOptions
+  ): Promise<RegisterSchemaReturn> {
+    //TODO: Handle validations
+    try {
+      const cheqdDidRegistrar = agentContext.dependencyManager.resolve(CheqdDidRegistrar)
+
+      const schema = options.schema
+      const schemaResource = {
+        id: v4(),
+        name: schema.name,
+        resourceType: 'anonCredsSchema',
+        data: {
+          name: schema.name,
+          version: schema.version,
+          attrNames: schema.attrNames,
+        },
+        version: schema.version,
+      } as CheqdCreateResourceOptions
+
+      const response = await cheqdDidRegistrar.createResource(agentContext, schema.issuerId, schemaResource)
+      if (response.resourceState.state !== 'finished') {
+        throw new Error(response.resourceState.reason)
+      }
+
+      return {
+        schemaState: {
+          state: 'finished',
+          schema,
+          schemaId: `${schema.issuerId}/resources/${schemaResource.id!}`,
```

Comment:
Nit, but if we generate the id outside of the resource it can never be undefined and we don't have to use `!`